customer/src/main/java/com/yas/customer/service/UserAddressService.java

```diff
@@ -1,46 +1,95 @@
+import com.yas.customer.viewmodel.address.AddressPostVm;
+import com.yas.customer.viewmodel.user_address.UserAddressVm;
+import org.springframework.data.domain.*;
 import org.springframework.security.core.context.SecurityContextHolder;
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
 
 import java.util.ArrayList;
+import java.util.Comparator;
 import java.util.List;
 
 @Service
 @Transactional
-@RequiredArgsConstructor
 public class UserAddressService {
     private final UserAddressRepository userAddressRepository;
+    private final LocationService locationService;
 
-    public List<AddressGetVm> getUserAddressList() {
+    public UserAddressService(UserAddressRepository userAddressRepository, LocationService locationService) {
+        this.userAddressRepository = userAddressRepository;
+        this.locationService = locationService;
+    }
+
+    public AddressListVm getUserAddressList(int pageNo, int pageSize) {
         String userId = SecurityContextHolder.getContext().getAuthentication().getName();
+        if(userId.equals("anonymousUser"))
+            throw new AccessDeniedException("Please login");
 
         List<UserAddress> userAddressList = userAddressRepository.findAllByUserId(userId);
-        return userAddressList.stream()
-                .map(userAddress -> {
-                    AddressGetVm addressGetListResponseVm = new AddressGetVm(
-                            userAddress.getId(), userAddress.getIsActive());
-                    return addressGetListResponseVm;
-                })
-                .toList();
+        List<AddressVm> addressVmList = locationService.getAddressesByIdList(userAddressList.stream()
+                .map(userAddress -> userAddress.getAddressId()).toList());
+
+        List<ActiveAddressVm> addressActiveVms = new ArrayList<>();
+        for (UserAddress userAddress : userAddressList) {
+            for (AddressVm addressVm : addressVmList) {
+                if (userAddress.getAddressId().equals(addressVm.id())) {
+                    addressActiveVms.add(new ActiveAddressVm(
+                            addressVm.id(),
+                            addressVm.contactName(),
+                            addressVm.phone(),
+                            addressVm.addressLine1(),
+                            addressVm.city(),
+                            addressVm.zipCode(),
+                            addressVm.districtId(),
+                            addressVm.stateOrProvinceId(),
+                            addressVm.countryId(),
+                            userAddress.getIsActive()
+                    ));
+                    //remove element to reduce the number of iterations
+                    addressVmList.remove(addressVm);
+                    break;
+                }
+            }
+        }
+
+        //handle pageable
+        Sort sort = Sort.by("isActive").descending();
```

Comment:
It look like pagination is done in memory, which is a bad practice.
Normally a user doesn't have too many address, so pagination might not need here.

Reply:
I'll update, thanks